types/src/account_state_view.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    access_path::AccessPath,
+    account_config::{AccountResource, BalanceResource, CRSNResource, ChainIdResource},
+    on_chain_config::{
+        access_path_for_config, dpn_access_path_for_config, ConfigurationResource, OnChainConfig,
+        ValidatorSet, Version,
+    },
+    state_store::state_key::StateKey,
+    validator_config::ValidatorConfig,
+};
+use move_core_types::{
+    account_address, account_address::AccountAddress, move_resource::MoveResource,
+};
+use serde::de::DeserializeOwned;
+
+pub struct AccountStateView<'a, F>
```

Comment:
Great idea to introduce this. 

(I'd just call it `AccountView` to distinguish from `StateView` and be shorter.)

Again, try to create a `trait AsAccountView` and impl it for all `S: StateView`, which will simplify the call sites.

Reply:
Introduce AccountViewTrait? 

```
trait AccountViewTrait {
    fn get();

    fn get_balance() {
        ... // blanket impl that calls self.get()
    }

    ...
}

impl AccountViewTrait for HashMap<.., ..>

impl AccountViewTrait for AccountState<.., ..>

impl AccountViewTrait for AccountView

struct AccountView<S: StateView> {
    address: AccountAddress,
    state_view: S
}
```

Reply:
Hmm but that would mean a lot of code duplication as the actual implementations need to rewrite similar functions?

Reply:
no the actual impl provide `get()` only.

Reply:
@msmouse - Thanks for the suggestion. I actually had started with this approach but resorted to changing it to the current one because I faced some issues. While re-implementing your suggestion, I now recall the issue I ran into. So the issue is for DbReader, we need two implementation of AccountViewTrait one with version and other without version. AFAIK, rust won't allow something like

```
impl AccountView for (dyn DbReader, Version) {
```

It doesn't like trait implementation for arbitrary types. So one way to do that is create one more wrapper type with db_reader and version. The same issue exists for HashMap as well - we need to wrap the hashmap in another type to be able to define a trait implementation. I think this becomes less ergonomic as compared to current approach which is provides flexibility of taking arbitrary fn pointer. Let me know if you think defining the wrapper is a better way, we can do that.

Reply:
```
impl<T: DbReader> AccountViewTrait for (T, Version) {
}
```

^ you implement for arbitrary dbreaders like this.. but I do prefer the wrapper, (around `S: StateView`, not the dbreader.).


For things that are not a `StateView`, namely `AccountState` and the  `HashMap` we implement the trait directly for them.

Reply:
>>For things that are not a StateView, namely AccountState and the HashMap we implement the trait directly for them.

We can't implement these traits directly on HashMap, I have defined a separate struct for doing that.